connectd/tor.c

```diff
@@ -57,29 +98,30 @@ static struct io_plan *connect_finish(struct io_conn *conn,
 	status_io(LOG_IO_IN, NULL, "proxy",
 		  connect->buffer, SIZE_OF_IPV4_RESPONSE + SIZE_OF_RESPONSE);
 
-	if ( connect->buffer[1] == '\0') {
-		if ( connect->buffer[3] == SOCKS_TYP_IPV6) {
-			/* Read rest of response */
-			return io_read(conn,
-				       connect->buffer + SIZE_OF_RESPONSE +
-				       SIZE_OF_IPV4_RESPONSE,
-				       SIZE_OF_IPV6_RESPONSE -
-				       SIZE_OF_IPV4_RESPONSE,
-				       &connect_finish2, connect);
-
-		} else if ( connect->buffer[3] == SOCKS_TYP_IPV4) {
-			status_debug("Now try LN connect out for host %s",
-				     connect->host);
-			return connection_out(conn, connect->connect);
-		} else {
-			status_debug
-			    ("Tor connect out for host %s error invalid type return ",
+	const char reply_status_code = connect->buffer[1];
+
+	if ( reply_status_code != SOCKS5_REPLY_SUCCESS) {
+		status_debug("Error connecting to %s: Tor server reply: %s",
+			     connect->host, socks5strerror(reply_status_code));
+		return io_close(conn);
+	}
+
+	if ( connect->buffer[3] == SOCKS_TYP_IPV6) {
+		/* Read rest of response */
+		return io_read(conn,
+			       connect->buffer + SIZE_OF_RESPONSE +
+			       SIZE_OF_IPV4_RESPONSE,
+			       SIZE_OF_IPV6_RESPONSE -
+			       SIZE_OF_IPV4_RESPONSE,
+			       &connect_finish2, connect);
+
+	} else if ( connect->buffer[3] == SOCKS_TYP_IPV4) {
+		status_debug("Now try LN connect out for host %s",
 			     connect->host);
-			return io_close(conn);
-		}
+		return connection_out(conn, connect->connect);
 	} else {
-		status_debug("Tor connect out for host %s error: %x ",
-			     connect->host, connect->buffer[1]);
+		status_debug("Tor connect out for host %s error invalid type "
+			     "return: %0x", connect->host, connect->buffer[3]);
 		return io_close(conn);
 	}
```

Comment:
What about address type "DOMAINNAME (03)" here? It is not implemented and I left it as is, but
https://tools.ietf.org/html/rfc1928#section-6
says `connect->buffer[3]` could also be `03`:
>             o  IP V4 address: X'01'
>             o  DOMAINNAME: X'03'
>             o  IP V6 address: X'04'

I think the code should just consume it an ignore it (like it does with IPv4 and IPv4 types) to be fully compliant with the Socks5 spec.